### PR TITLE
80 dataset drops (also 82, 83, 90)

### LINKED ISSUE #80: Dataset Manipulation
Following our meeting w/ partners, we need to be able to:

- Randomly drop observations from datasets
- Create datasets containing only 2 (or some other number) classes

### LINKED ISSUE #82: Refactor metric experiment for config objects
The `MetricExperiment` class was implemented prior to the `ConfigClass` approach being taken up, and still works by taking a dictionary as an input. This should be refactored to be in line with the current approach.

### LINKED ISSUE #83: Dataset train/val/test splits
Check implementation and ensure training & metrics are using the same samples etc.

### LINKED ISSUE #90: `wandb_args` on lower-level configs but `wandb` on top level configs
Minor comment, but is there a reason for calling this `wandb_args` on lower-level configs but `wandb` on top level configs? I have a preference for consistency across these argument names to minimise confusion

_Originally posted by @philswatton in https://github.com/alan-turing-institute/ARC-LoCoMoSeT/pull/76#discussion_r1405903272_

### PR DESCRIPTION
**Pull request by philswatton**
This PR adds several functions, mainly in src/locomoset/data/preprocess, to:

- Resolves #80 by adding two functions, `drop_images` and `drop_images_by_label`
  - `drop_images` takes `keep_size` as an argument, which can be either an int or a float between 0 and 1 specifying how many images to keep. The drops are stratified by label so pre-existing label balance will be preserved
  - `drop_images_by_label` takes `keep_labels` as an argument, which is a list containing all labels to keep. All images without these labels are dropped.
  - `apply_dataset_mutations` takes both arguments and applies both functions (dropping by label first, randomly dropping images second)
  - Under the hood, these functions use `_mutate_dataset` which takes a dataset or dataset dict, a function, and kwargs to that function, then applies it. This makes it easier to ignore what kind of object is being worked with
- Resolves #82 by refactoring the experiment to use a config class, rather than a dict
- Resolves #83 by adding train/val/test splits via `create_data_splits`. This will automatically detect whether a single dataset is being split into three, a dict of two datasets needs val or test to be added (i.e. train must always exist - the val or test set will be created from train), or if it is already split three ways, and return a dataset dict with all three splits
  - Note this is a change in behaviour: `train_split` and `val_split` (and by extension, the new `test_split` should NOT contain the same names. If these splits exist, they should match the names in the dataset: otherwise they are used to create new splits with the corresponding names
- Resolves #90 by ensuring top-level configs are consistent and use `wandb_args`, not `wandb`
- Adds tests for all new functions

This involved some refactoring in places, as running training and metrics experiments via configs needed to be rewritten. Given the same arguments in a config (in terms of dataset name, split names, val/test sizes), both training and metrics will use

Unit tests have been updated in some places where relevant

Not implemented in this PR:
- #81: the configs are a *little* less permissive where I've been changing stuff, but I've left actually doing this for a PR in the new year
- actually using the test set

### CODE REVIEW

**Comment by jack89roberts** [comment]
Keeping track of a few things here:

- I refactored the splitting a bit, mostly so that we keep a large test dataset in all cases (i.e. don't scale it down along with the train/val dataset sizes). The size of the `train` split is now controlled by `n_samples` (as used by the metrics). The `val` split is further subsetted before training starts to make it at most one-quarter of the training dataset size - best to keep it small since it's only for debugging whether the training has gone ok now.

- I've made it so the training jobs log the test dataset accuracy (once at the end).

- The LoCoMoSeT metrics will be computed on the train dataset only now, rather than train + val. I felt it was weird to potentially be computing the metrics on more (or different) images than the model sees in training. Could be convinced to revert that.

- **_[Should only an issue for sub-setting existing datasets, should not be an issue for the binary datasets created elsewhere]_** After dropping labels the `ClassLabel` in the dataset (`dataset["train"].features["label"]`) still has all the original labels and the class IDs correspond to the original dataset (so e.g. they may be 7 and 23 instead of 0 and 1). This means we will end up training a model with e.g. a 100 class classification head but it will only see 2 of those classes during training (and the metrics will only be computed on those classes). Subsetting a `ClassLabel` instance and updating the dataset labels to reflect isn't a built-in feature of HF datasets as far as I could tell.

- `keep_labels` is not nested under `dataset_args` in the top level configs but is in the individual configs (not attempted to fix yet, partially since `keep_labels` is set to be one of the things that can be looped over and unsure how messy that would get).

**Comment by jack89roberts** [comment]
Going to merge this so we don't have a messy mix of PRs but can talk through the above when you're back @philswatton

**Review by jack89roberts — APPROVED** [review]
